src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -4,44 +4,257 @@
 
 using System;
 using System.Collections.Generic;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
 
 using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
 
 namespace ILCompiler
 {
     internal class ReadyToRunMetadataFieldLayoutAlgorithm : MetadataFieldLayoutAlgorithm
     {
         /// <summary>
-        /// CoreCLR DomainLocalModule::OffsetOfDataBlob() / sizeof(void *)
+        /// Map from EcmaModule instances to field layouts within the individual modules.
         /// </summary>
-        private const int DomainLocalModuleDataBlobOffsetAsIntPtrCount = 6;
+        private ModuleFieldLayoutMap _moduleFieldLayoutMap;
 
-        /// <summary>
-        /// CoreCLR ThreadLocalModule::OffsetOfDataBlob() / sizeof(void *)
-        /// </summary>
-        private const int ThreadLocalModuleDataBlobOffsetAsIntPtrCount = 3;
-
-        private LayoutInt _initialNonGcStaticsOffset;
+        public ReadyToRunMetadataFieldLayoutAlgorithm(TargetDetails target)
+        {
+            _moduleFieldLayoutMap = new ModuleFieldLayoutMap(target);
+        }
 
-        public ReadyToRunMetadataFieldLayoutAlgorithm(TargetDetails target, int numberOfTypesInModule)
+        public override ComputedStaticFieldLayout ComputeStaticFieldLayout(DefType defType, StaticLayoutKind layoutKind)
         {
-            _initialNonGcStaticsOffset = new LayoutInt(DomainLocalModuleDataBlobOffsetAsIntPtrCount * target.PointerSize + numberOfTypesInModule);
+            ComputedStaticFieldLayout layout = new ComputedStaticFieldLayout();
+            if (defType is EcmaType ecmaType)
+            {
+                // ECMA types are the only ones that can have statics
```

Comment:
We'll probably have to revisit this comment when we look at statics on generic types.

Reply:
I guess we'll need to do something else, the layout algorithm I just implemented explicitly skips generics.